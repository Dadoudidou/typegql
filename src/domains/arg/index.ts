import { argRegistry } from './registry';
export { compileFieldArgs } from './compiler';
import { ArgOptions, defaultArgOptions } from './options';

export function Arg(options: ArgOptions = {}): ParameterDecorator {
  return (target: Object, fieldName: string, argIndex: number) => {
    const compiledOptions = {
      ...defaultArgOptions,
      ...options,
    }
    argRegistry.set(target.constructor, [fieldName, argIndex], compiledOptions);
  };
}
