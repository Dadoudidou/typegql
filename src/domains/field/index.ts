import { fieldsRegistry, FieldInnerConfig } from './registry';

export { FieldInnerConfig, fieldsRegistry, queryFieldsRegistry } from './registry';
export { compileAllFields, compileFieldConfig } from './compiler';
export { isQueryField } from './services';
export { FieldError } from './error';

export interface FieldOptions {
  description?: string;
  type?: any;
  name?: string;
  isNullable?: boolean;
  isList?: boolean;
}

export function Field(options?: FieldOptions): PropertyDecorator {
  return (targetInstance: Object, fieldName: string) => {
    const finalConfig: FieldInnerConfig = {
      property: fieldName,
      name: fieldName,
      isList: false,
      isNullable: true,
      ...options,
    };

    fieldsRegistry.set(targetInstance.constructor, fieldName, {
      ...finalConfig,
    });
  };
}
