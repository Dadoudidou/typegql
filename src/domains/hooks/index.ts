import {
  registerFieldAfterHook,
  registerFieldBeforeHook,
  HookExecutor,
} from './registry';

export {
  fieldAfterHooksRegistry,
  fieldBeforeHooksRegistry,
  HookExecutor,
} from './registry';
export { HookError } from './error';

export function Before(hook: HookExecutor): PropertyDecorator {
  return (targetInstance: Object, fieldName: string) => {
    registerFieldBeforeHook(targetInstance.constructor, fieldName, hook);
  };
}

export function After(hook: HookExecutor): PropertyDecorator {
  return (targetInstance: Object, fieldName: string) => {
    registerFieldAfterHook(targetInstance.constructor, fieldName, hook);
  };
}

export interface GuardOptions {
  msg: string;
}

export function Guard(
  guardFunction: HookExecutor<boolean>,
  options: GuardOptions,
): PropertyDecorator {
  return (targetInstance: Object, fieldName: string) => {
    const hook: HookExecutor = async data => {
      const isAllowed = await guardFunction(data);
      if (isAllowed !== true) {
        throw new Error(options.msg);
      }
    };
    registerFieldBeforeHook(targetInstance.constructor, fieldName, hook);
  };
}

export function createGuard(guardFunction: HookExecutor<boolean>, options: GuardOptions) {
  return (guardOptions: GuardOptions = options): PropertyDecorator => {
    return Guard(guardFunction, guardOptions);
  };
}
