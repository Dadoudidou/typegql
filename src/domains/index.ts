export { Type, compileType } from './type';
export { Field } from './field';
export { Arg } from './arg';
export { Inject, Context, Source, Info } from './inject';
export { registerEnum } from './enum';
