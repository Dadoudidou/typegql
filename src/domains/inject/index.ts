import { InjectorResolver, injectorRegistry } from './registry';
export { injectorRegistry, InjectorsIndex, InjectorResolver } from './registry';

export function Inject(resolver: InjectorResolver): ParameterDecorator {
  return (target: Object, fieldName: string, argIndex: number) => {
    injectorRegistry.set(target.constructor, [fieldName, argIndex], resolver);
  };
}

export const Context = Inject((source, args, context, info) => {
  return context;
});

export const Info = Inject((source, args, context, info) => {
  return info;
});

export const Source = Inject((source, args, context, info) => {
  return source;
});
