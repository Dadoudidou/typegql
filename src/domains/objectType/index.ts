import { compileObjectTypeWithConfig, compileObjectTypeAsInputWithConfig } from './compiler';
import { objectTypeRegistry, inputTypeRegistry } from './registry';

export { compileObjectType } from './compiler';
export { ObjectTypeError } from './error';
export { objectTypeRegistry, inputTypeRegistry } from './registry';

export interface ObjectTypeOptions {
  name?: string;
  description?: string;
}

export function ObjectType(options?: ObjectTypeOptions): ClassDecorator {
  return (target: Function) => {
    const config = { name: target.name, ...options };
    const outputTypeCompiler = () =>
      compileObjectTypeWithConfig(target, config);
    objectTypeRegistry.set(target, outputTypeCompiler);
    
    const inputTypeCompiler = () =>
    compileObjectTypeAsInputWithConfig(target, config);
    inputTypeRegistry.set(target, inputTypeCompiler);
  };
}
