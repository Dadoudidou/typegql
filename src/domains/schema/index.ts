export { schemaRegistry, mutationFieldsRegistry, queryFieldsRegistry } from './registry';
import { schemaRegistry } from './registry';
import { compileSchema } from './compiler';
export { compileSchema } from './compiler';
export { Query, Mutation } from './rootFields';

export function Schema(): ClassDecorator {
  return target => {
    const compiler = () => compileSchema(target);
    schemaRegistry.set(target, compiler);
  };
}
