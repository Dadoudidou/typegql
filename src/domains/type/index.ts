import { registerType } from './registry';

export { objectTypeRegistry, inputTypeRegistry, registerType } from './registry';
export { ObjectTypeError } from './error'

export interface ObjectTypeOptions {
  name?: string;
  description?: string;
}

export function ObjectType(options?: ObjectTypeOptions): ClassDecorator {
  return (target: Function) => {
    const config = { name: target.name, ...options };
    registerType(target, config);
  };
}
