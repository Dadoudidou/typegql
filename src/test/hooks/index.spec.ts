import { ObjectType, Field, Before, After, Guard, compileObjectType } from 'domains';

describe('Hooks', () => {
  it('Will call @Before hook on field resolve', async () => {
    const callback = jest.fn();
    @ObjectType()
    class Foo {
      @Field()
      @Before(callback)
      bar: string = 'done';
    }

    const { bar } = compileObjectType(Foo).getFields();

    await bar.resolve(null, null, null, null);

    expect(callback).toBeCalled();
  });

  it('Will call @After hook on field resolve', async () => {
    const callback = jest.fn();
    @ObjectType()
    class Foo {
      @Field()
      @After(callback)
      bar: string = 'done';
    }

    const { bar } = compileObjectType(Foo).getFields();

    await bar.resolve(null, null, null, null);

    expect(callback).toBeCalled();
  });

  it('Will call @Before cb first, then resolver, and then @After cb', async () => {
    const beforeCb = jest.fn();
    const resolverCb = jest.fn();
    const afterCb = jest.fn();

    let counter: number = 0;

    const incrementAndCall = (cb: Function) => () => {
      counter++;
      cb(counter);
    };

    @ObjectType()
    class Foo {
      @Field()
      @Before(incrementAndCall(beforeCb))
      @After(incrementAndCall(afterCb))
      bar(): string {
        incrementAndCall(resolverCb)();
        return 'done';
      }
    }

    const { bar } = compileObjectType(Foo).getFields();

    const result = await bar.resolve(null, null, null, null);

    expect(beforeCb).toBeCalledWith(1);
    expect(resolverCb).toBeCalledWith(2);
    expect(afterCb).toBeCalledWith(3);
    expect(result).toEqual('done');
  });

  it('Will not allow resolver to execute if @Guard returns false with error message', async () => {
    const callback = jest.fn();
    @ObjectType()
    class Foo {
      @Guard(
        (source, args) => {
          return args.isAllowed;
        },
        { msg: 'Not allowed' },
      )
      @Field()
      bar(isAllowed: boolean): string {
        callback();
        return 'done';
      }
    }

    const { bar } = compileObjectType(Foo).getFields();

    await expect(
      bar.resolve(null, { isAllowed: false }, null, null),
    ).rejects.toThrowErrorMatchingSnapshot();
    expect(callback).not.toBeCalled();
    await expect(bar.resolve(null, { isAllowed: true }, null, null)).resolves.toEqual(
      'done',
    );
    expect(callback).toBeCalled();
  });
});
