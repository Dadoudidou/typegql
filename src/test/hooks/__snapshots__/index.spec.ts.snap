// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Hooks Will not allow resolver to execute if @Guard returns false with error message 1`] = `"Not allowed"`;

exports[`Hooks Will work with createGuard @Guards 1`] = `"Not allowed here"`;
